third_party/amd/lib/TritonAMDGPUTransforms/RescheduleOps.cpp

```diff
@@ -225,368 +371,1164 @@ struct Graph {
+            std::to_string(reinterpret_cast<intptr_t>(dep.parent));
+        std::string childAddr =
+            std::to_string(reinterpret_cast<intptr_t>(dep.child));
+        out << "\t" << childAddr << " -> " << parentAddr << " [color=" << color
+            << ", style=" << style << "]\n";
+      }
+    }
+    out << "}\n";
+    return out;
+  }
+
+private:
+  BasicBlockNodeMap nodeMap; // does not get changed; contains shared_ptrs.
+  SchedDagNodeList nodeList; // rewritten every rescheduling.
+  SchedDag dag;
+  bool readyToAddDeps;
+  DepMap deps;
+  SmallVector<StringRef> depOrder;
+  DenseMap<SchedDagNode *, size_t> nodeIndices;
+  int32_t rescheduleId;
+};
+
+/******************************************************************************
+  TritonAMDGPURescheduleOps::applyReschedulingPasses()
+  is the top-level scheduling pass for a single block,
+  whose purpose is to improve performance and regalloc of backend compilers.
+  Before this pass, mfmas and local_loads (belonging to dots)
+  were already annotated with their dot-tile info.
+  The order of re-scheduling is:
+    - Place order dependencies on dots according to dot-tiling.
+    - Place order dependencies on local_loads according to dot-tiling.
+    - Determine min-register vs max-latency-hiding preference.
+    - Determine memory op order and co-scheduling.
+    - Place order dependencies between memory ops.
+    - Determine memory ops' early/late preference.
+    - Determine memory ops' preferred issue rate.
+    - Determine memory ops' supported issue rate.
+    - Place performance and anti-dependencies between memory ops and dots.
+    - Run scheduler with new dependencies in place.
+  Note that reschedule() can be run after any new dependencies are created to
+  visualize dag.
+******************************************************************************/
+struct TritonAMDGPURescheduleOps
+    : public TritonAMDGPURescheduleOpsBase<TritonAMDGPURescheduleOps> {
+  explicit TritonAMDGPURescheduleOps(StringRef targetArch) {
+    this->arch = targetArch.str();
+  }
 
-    std::string outStr;
-    llvm::raw_string_ostream outStream(outStr);
-    outStream << "\n\n\n...." << std::string(80, '-') << '\n';
-    for (auto op : rescheduledOps) {
-      op->print(outStream);
-      outStream << "\n";
+  LogicalResult verify(Block *mlirBlock) {
+    // make sure that a block gets terminated with `cf::BranchOp`
+    if (!dyn_cast<mlir::cf::BranchOp>(&(mlirBlock->back()))) {
+      return failure();
     }
 
-    // re-order instruction based on the new schedule
-    // move instruction from the tail to the begining of the current BB
-    // one-by-one
+    // don't schedule if there is not enough operations in a block
+    if (mlirBlock->getOperations().size() < 3)
+      return failure();
+    return success();
+  }
+
+  void applyReschedulingPasses(Block *mlirBlock) {
+    LDBG("");
+    LDBG("");
+    llvm::outs() << "TritonAMDGPURescheduleOps::applyReschedulingPasses()\n";
```

Comment:
move it to `LDBG` - e.g., `LDBG("TritonAMDGPURescheduleOps::applyReschedulingPasses()");

Let's use `llvm::outs()` if some user-provided `verbose` flag is passed